.github/workflows/build-artifact.yaml

```diff
@@ -54,14 +65,42 @@ jobs:
       - name: Set up dynamic env vars for build
         run: |
           echo "TT_METAL_HOME=$(pwd)" >> $GITHUB_ENV
+          echo "RUNNER_UID=$(id -u)" >> $GITHUB_ENV
+          echo "RUNNER_GID=$(id -g)" >> $GITHUB_ENV
       - name: Update submodules
         run: |
           git submodule update --init --recursive
-      - name: Build tt-metal and libs
+      - name: Determine Docker Image Tag
         run: |
-          nice -n 19 cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=${{ inputs.build-type }} -DENABLE_TRACY=${{ inputs.tracy }}
-          nice -n 19 cmake --build build --target tests
-          nice -n 19 cmake --build build --target install
+          if [[ "${GITHUB_REF_NAME}" == "main" ]]; then
+            echo "IMAGE_TAG=latest" >> $GITHUB_ENV
+          else
+            echo "IMAGE_TAG=dev-${GITHUB_REF_NAME//\//-}" >> $GITHUB_ENV
+          fi
+      - name: Docker login
+        uses: docker/login-action@v3
+        with:
+          registry: https://ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build tt-metal and libs
+        uses: addnab/docker-run-action@v3
+        with:
+          image: ghcr.io/tenstorrent/tt-metal/tt-metalium/${{ inputs.os }}:${{ env.IMAGE_TAG }}
```

Comment:
So let me make sure I have this clear:

- For `build-artifact` right now, given a certain branch, we will ALWAYS use the latest version of the image on that branch.
- This means that, let's say, a perf pipeline runs at 2am and generates a new image `Main-A`. For attempt #1 of that pipeline, it will use that image `Main-A`.
- Let's say at 10am, after a couple of commits to main, the latest image is now `Main-B`. Let's say people have made some changes to the image by then.
- Let's say the perf pipeline at 2am failed. I wake up, and decide to re-run it. However, since it's on main, it will use latest image, which is now `Main-B` image instead of `Main-A`. This means that Attempt #1 uses a different image from Attempt #2.

Could this not potentially cause inconsistency / instability?

Reply:
Now of course, since for this PR it only concerns build, if the silicon test step failed, it should already use build artifacts generated at 2am, assuming that build artifact passed.

Let's just say for this mental exercise though. I'm either retrying the entire pipeline start to finish OR the build artifact somehow did fail. Let's say a network error.

Reply:
to answer your question point by point
we will always build a new docker-image based on the branch name. if it is on main branch, it will point to `latest`, else it will default to `dev-BRANCH`. This means that for any single branch, there is a docker image tag associated with it. 

The assumptions here are that for tests run on main branch, we assume that deps introduced are vetted before merged (as it should be).

In your example, if the perf pipeline failed on `Main-A` and on `Main-B`, that would indicate the new changes to the system did not affect the failures on Main-A, no?

in any case, in our current system, if we added new deps and installed them on our CI machines manually, we would be in the same situation so I do not think this would be a docker specific issue. In fact, adding docker would mean we can identify which deps were changed and easily revert those changes (if we so wished to)

Reply:
I think I'm worried about the case where for example, `Main-A` has a some dependency `depA` that could no longer exist in `Main-B`. Perhaps the perf pipelines at the point of `Main-B` no longer needs `depA` (hopefully that's what you mean by vetted), but for sure `Main-A` still will.

If we retry the perf pipeline at `Main-A`, but use the latest image from `Main-B` or onward, it would no longer have the needed `depA`.

This would affect for example, the package and release pipeline.